stetho/src/main/java/com/facebook/stetho/inspector/elements/Document.java

```diff
@@ -0,0 +1,602 @@
+import com.facebook.stetho.common.ArrayListAccumulator;
+import com.facebook.stetho.common.LogUtil;
+import com.facebook.stetho.common.ThreadBound;
+import com.facebook.stetho.common.UncheckedCallable;
+import com.facebook.stetho.inspector.helper.ObjectIdMapper;
+import com.facebook.stetho.inspector.helper.ThreadBoundProxy;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Queue;
+import java.util.regex.Pattern;
+
+public final class Document extends ThreadBoundProxy {
+  private final DocumentProvider.Factory mFactory;
+
+  private DocumentProvider mDocumentProvider;
+  private ShadowDocument mShadowDOM;
+  private ObjectIdMapper mObjectIdMapper;
+
+  private UpdateListenerCollection mUpdateListeners;
+
+  private ChildEventingList mCachedChildEventingList;
+  private ArrayListAccumulator<Object> mCachedChildrenAccumulator;
+
+  private int mReferenceCounter;
+
+  public Document(DocumentProvider.Factory factory) {
+    super(factory);
+
+    mFactory = factory;
+    mObjectIdMapper = new DOMObjectIdMapper();
+    mReferenceCounter = 0;
+    mUpdateListeners = new UpdateListenerCollection();
+  }
+
+  public synchronized void addRef() {
+    if (mReferenceCounter++ == 0) {
+      mDocumentProvider = mFactory.create();
+    }
+  }
+
+  public synchronized void release() {
+    if (mReferenceCounter > 0) {
+      if (--mReferenceCounter == 0) {
+        cleanUp();
+      }
+    }
+  }
+
+  private void cleanUp() {
+    mDocumentProvider.postAndWait(new Runnable() {
+      @Override
+      public void run() {
+        mDocumentProvider.setListener(null);
+        mShadowDOM = null;
+        mObjectIdMapper.clear();
+        mDocumentProvider.dispose();
+        mUpdateListeners.clear();
+        mDocumentProvider = null;
+      }
+    });
+  }
+
+  public void addUpdateListener(UpdateListener updateListener) {
+    mUpdateListeners.add(updateListener);
+  }
+
+  public void removeUpdateListener(UpdateListener updateListener) {
+    mUpdateListeners.remove(updateListener);
+  }
+
+  public NodeDescriptor getNodeDescriptor(Object element) {
+    verifyThreadAccess();
+    return mDocumentProvider.getNodeDescriptor(element);
+  }
+
+  public void highlightElement(Object element, int color) {
+    verifyThreadAccess();
+    mDocumentProvider.highlightElement(element, color);
+  }
+
+  public void hideHighlight() {
+    verifyThreadAccess();
+    mDocumentProvider.hideHighlight();
+  }
+
+  public void setInspectModeEnabled(boolean enabled) {
+    verifyThreadAccess();
+    mDocumentProvider.setInspectModeEnabled(enabled);
+  }
+
+  public int getNodeIdForElement(Object element) {
+    verifyThreadAccess();
+    return mObjectIdMapper.getIdForObject(element);
+  }
+
+  public Object getElementForNodeId(int id) {
+    verifyThreadAccess();
+    return mObjectIdMapper.getObjectForId(id);
+  }
+
+  public DocumentView getDocumentView() {
```

Comment:
One thing we're going to have to be careful with here is the fact that `DocumentView` seems like it should be providing an immutable "view". However, it's not. `ShadowDOM` can and will change if you access it across tree updates.

There are a few things we should do here. First, we should add documentation to this method that says you should only use the `DocumentView` "temporarily" ... don't stash it in a field and access it later and expect it to be unchanged! Only use it for the duration of the calling method. Second, let's file a task to go and make this area less prone to interpretation errors (it's a potential bug hazard). That may mean we rename some things, or it may mean that `ShadowDOM` doesn't implement `DocumentView` directly and instead can hand out a `DocumentView` which is, in fact, immutable.